0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+in some Dafny codebases.
+This also introduces the same proof obligations as the previous solution,
+although here they could be proved by numerous helper methods that operate on the subset type
+in the same shared library.
+
+This would likely be the best solution if changing the Dafny language itself was not an option.
+The major downside is that built-in sequence operations such as `s[i]` and `s + s'` would still require additional
+effort to achieve verification, or would have to be abandoned entirely in favour of the helper methods
+provided by the shared library.
+
+## Change the definition of the `string` type
+
+The current definition of `string` as an alias for `seq<char>` is problematic because surrogate code points
+are not truly independent values in the sequence.
+It is actually quite unusual for a general-purpose language to treat strings as only a special case of
+generic datatypes for arbitrary sequences of values.
+Dafny could instead make `string` an alias for a different type, whether built-in or defined in Dafny source in the runtime or a standard library,
+that treats strings as a more specialized datatype. It could then be more prescriptive about picking a definitive encoding to
+support directly (most likely UTF-8).
+
+This would certainly require more effort for existing code to migrate to.
+It is also a much more expensive solution to implement, since it requires creating an API for strings
+that is at least adequate to replace all existing usage of sequence operations.
+These interfaces often become very large to support all common string operations efficiently,
+and would need very careful thought.
+Although unusual, representing strings directly as sequences of abstract characters
+works very well for a verification-focussed language like Dafny,
+since sequences are already a deep built-in concept in the language semantics.
+
+This new type could alternatively be introduced with a different name, such as `unicode` as in Python 2,
+while keeping the alias of `string` for `seq<char>`.
+This would only increase the confusion and cognitive burden for Dafny users in the future, though.
+
+## Add a new, distinct `rune` type
+
+We could maintain the current definition of the `char` type, and introduce a new `rune` type to represent Unicode scalar values
+instead ("rune" being the term both Go and C# use for this).
+This would make it more obvious when reading Dafny source code in isolation whether it was using the new definitions of strings.
+There are often few explicit references to individual character values in code that references string values, however,
+and even when the `char` type is used it is often only implicitly referenced because of type inference.
+This alternative does not seem to be worth the implementation cost or additional cognitive load on users,
+especially since it is hard to imagine any codebase needing to use both `char` and `rune` together.
```

Comment:
I'm not sure about this (never using both char and rune): IIRC Windows file system APIs use sequences of Dafny-3 `char`s, not valid utf-16 strings.  

Maybe this just means that we'll have to build file system APIs with `seq<byte>` instead of `string`.

Reply:
Yeah, if you need to manipulate possibly-invalid sequences of UTF-16 code points, that should probably be a `seq<uint16>` instead.

I'm personally happy with the trade-off of a simpler, maximally correct and portable definition of `string` at the expense of making some specific APIs more ugly.

Reply:
Added an explicit note on this, and an open question on being able to compile to a native utf-16 type.